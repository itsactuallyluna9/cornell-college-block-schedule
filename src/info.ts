import loadCalendar from "./calendar.ts";

export type CalendarData = {
  is_weekend: boolean;
  block: {
    block: number;
    start: Date;
    end: Date;
  } | null;
  day: number | null;
  week: number | null;
  event: {
    name: string;
    start: Date;
    end: Date;
  } | null;
  school_year?: string;
};

export function calendarData(for_date: Date): CalendarData {
  const calendar = loadCalendar();
  const school_year = Object.keys(calendar).find((year) => {
    const year_data = calendar[year];
    const first_block = year_data.blocks[0];
    const last_block = year_data.blocks[year_data.blocks.length - 1];
    return for_date >= first_block.start && for_date <= last_block.end;
  });
  if (!school_year) {
    throw new Error("Calendar data not found for date!");
  }
  const year_data = calendar[school_year];
  const block = year_data.blocks.find((block) => {
    return for_date >= block.start && for_date <= block.end;
  });
  const event = year_data.events.find((event) => {
    return for_date >= event.start && for_date <= event.end;
  });

  if (block && event) {
    return {
      is_weekend: for_date.getDay() === 0 || for_date.getDay() === 6,
      block,
      day: null,
      week: null,
      event,
      school_year,
    };
  }

  if (!block && event) {
    return {
      is_weekend: for_date.getDay() === 0 || for_date.getDay() === 6,
      block: null,
      day: null,
      week: null,
      event,
      school_year,
    };
  }
  if (!block && !event) {
    // so if we're not in a block, and we're not in an offically defined break, we're in block break!
    // block break starts on the last day of the last block we were in, and ends on the first day of the next block
    // to get the last block we were in, we need to find the most recent block that ended before the date we're checking
    const last_block = year_data.blocks.reduce((prev, current) => {
      if (current.end < for_date) {
        return current;
      }
      return prev;
    }, [] as any);
    if (!last_block) {
      throw new Error("Calendar data not found for date!");
    }
    // now we need to find the next block!
    const next_block =
      year_data.blocks[year_data.blocks.indexOf(last_block) + 1];
    if (!next_block) {
      throw new Error("Calendar data not found for date!");
    }
    // if (next_block.block) isn't between 1-8
    if (next_block.block < 1 || next_block.block > 8) {
      // we're in pure summer!
      // summer break ends on the first day of the first block of the next school year

      // get the next school year
      const next_school_year = Object.keys(calendar).find((year) => {
        return year > school_year;
      });

      if (!next_school_year) {
        throw new Error("Calendar data not found for date!");
      }

      const next_year_data = calendar[next_school_year];
      const next_year_first_block = next_year_data.blocks[0];
      return {
        is_weekend: for_date.getDay() === 0 || for_date.getDay() === 6,
        block: null,
        day: null,
        week: null,
        event: {
          name: "Summer Break",
          start: last_block.end,
          end: next_year_first_block.start,
        },
      };
    }

    // now we can check if the date we're checking is between the last block and the next block
    if (for_date > last_block.end && for_date < next_block.start) {
      return {
        is_weekend: for_date.getDay() === 0 || for_date.getDay() === 6,
        block: null,
        day: null,
        week: null,
        event: {
          name: "Block Break",
          start: last_block.end,
          end: next_block.start,
        },
        school_year,
      };
    } else {
      throw new Error("Calendar data not found for date!");
    }
  }

  if (!block) {
    // make ts happy.
    throw new Error("Calendar data not found for date!");
  }

  const is_weekend = for_date.getDay() === 0 || for_date.getDay() === 6;
  // count up the days since the start of the block
  // skip weekends and events
  let day = 0;
  let week = 0;
  let had_event = false;
  const current_date = new Date(block.start);
  while (current_date <= for_date) {
    const event = year_data.events.find((event) => {
      return current_date >= event.start && current_date < event.end;
    });
    if (event) {
      had_event = true;
    }
    if (current_date.getDay() !== 0 && current_date.getDay() !== 6 && !event) {
      day++;
    }
    if (current_date.getDay() === 1) {
      week++;
    }
    current_date.setDate(current_date.getDate() + 1);
  }
  if (had_event) {
    day++;
    // and recalculate the week...
    week = Math.ceil(day / 5);
    if (is_weekend) {
      day--;
      week--;
    }
  }
  if (block.block < 1 || block.block > 8) {
    // we're in summer, but there's a block going on!
    // summer break ends on the first day of the first block of the next school year

    // get the next school year
    const next_school_year = Object.keys(calendar).find((year) => {
      return year > school_year;
    });

    if (!next_school_year) {
      throw new Error("Calendar data not found for date!");
    }
    const required_blocks = year_data.blocks.filter((block) => {
      return block.block >= 1 && block.block <= 8;
    });
    const end_of_last_required_block =
      required_blocks[required_blocks.length - 1].end;
    const next_year_data = calendar[next_school_year];
    const next_year_first_block = next_year_data.blocks[0];
    return {
      is_weekend: for_date.getDay() === 0 || for_date.getDay() === 6,
      block,
      day,
      week,
      event: {
        name: "Summer Break",
        start: end_of_last_required_block,
        end: next_year_first_block.start,
      },
    };
  } else {
    return {
      is_weekend,
      block,
      day,
      week,
      event: null,
      school_year,
    };
  }
}
