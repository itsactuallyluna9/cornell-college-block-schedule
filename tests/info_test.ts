import { test, expect } from "bun:test";
import { calendarData } from "../src/info.ts";

test("handles pure summer break", () => {
  const date_to_check = new Date("June 1, 2023 9:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();
});

test("handles not pure summer break", () => {
  const date_to_check = new Date("June 25, 2023 9:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();
});

test("handles block breaks", () => {
  const date_to_check = new Date("Feb 10, 2023 9:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();
});

test("handles actual blocks", () => {
  const date_to_check = new Date("Mar 3, 2023 9:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();
});

test("handles breaks in blocks", () => {
  const date_to_check = new Date("Nov 26, 2024 9:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();
});

test("handles breaks in blocks on return", () => {
  const date_to_check = new Date("Dec 2, 2024 8:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();

  const date_to_check_weekend = new Date("Dec 8, 2024 9:00");
  const data_weekend = calendarData(date_to_check_weekend);
  expect(data_weekend).toMatchSnapshot();
});

test("handles pure breaks", () => {
  const date_to_check = new Date("Dec 25, 2023 9:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();
});

test("throws on invalid school year", () => {
  expect(() => {
    calendarData(new Date("Jan 1, 1900 9:00"));
  }).toThrowError("Calendar data not found for date!");
  expect(() => {
    calendarData(new Date("Jan 1, 3000 9:00"));
  }).toThrowError("Calendar data not found for date!");
});

test("handles weekends", () => {
  const date_to_check = new Date("Nov 2, 2024 9:00");
  const data = calendarData(date_to_check);
  expect(data).toMatchSnapshot();
});
