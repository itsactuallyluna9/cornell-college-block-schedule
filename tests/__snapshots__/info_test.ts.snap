// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`handles pure summer break 1`] = `
{
  "block": null,
  "day": null,
  "event": {
    "end": 2023-08-28T14:00:00.000Z,
    "name": "Summer Break",
    "start": 2023-05-10T20:00:00.000Z,
  },
  "is_weekend": false,
  "week": null,
}
`;

exports[`handles not pure summer break 1`] = `
{
  "block": {
    "block": 9,
    "end": 2023-06-29T20:00:00.000Z,
    "start": 2023-06-05T14:00:00.000Z,
  },
  "day": 15,
  "event": {
    "end": 2023-08-28T14:00:00.000Z,
    "name": "Summer Break",
    "start": 2023-05-10T20:00:00.000Z,
  },
  "is_weekend": true,
  "week": 3,
}
`;

exports[`handles block breaks 1`] = `
{
  "block": null,
  "day": null,
  "event": {
    "end": 2023-02-13T15:00:00.000Z,
    "name": "Block Break",
    "start": 2023-02-08T21:00:00.000Z,
  },
  "is_weekend": false,
  "school_year": "2022-2023",
  "week": null,
}
`;

exports[`handles actual blocks 1`] = `
{
  "block": {
    "block": 6,
    "end": 2023-03-08T18:00:00.000Z,
    "start": 2023-02-13T15:00:00.000Z,
  },
  "day": 14,
  "event": null,
  "is_weekend": false,
  "school_year": "2022-2023",
  "week": 3,
}
`;

exports[`handles breaks in blocks 1`] = `
{
  "block": null,
  "day": null,
  "event": {
    "end": 2025-11-18T15:00:00.000Z,
    "name": "Block Break",
    "start": 2024-11-13T18:00:00.000Z,
  },
  "is_weekend": false,
  "school_year": "2024-2025",
  "week": null,
}
`;

exports[`handles breaks in blocks on return 1`] = `
{
  "block": null,
  "day": null,
  "event": {
    "end": 2025-11-18T15:00:00.000Z,
    "name": "Block Break",
    "start": 2024-11-13T18:00:00.000Z,
  },
  "is_weekend": false,
  "school_year": "2024-2025",
  "week": null,
}
`;
